windows/devices/uwb/UwbConnector.cxx

```diff
@@ -895,43 +868,256 @@ UwbConnector::NotificationListenerStop()
+{
+    std::lock_guard eventCallbacksLockExclusive{ m_eventCallbacksGate };
+    bool isFirstCallback = not CallbacksPresent();
+
+    auto OnSessionEndedToken = GetToken<::uwb::UwbRegisteredSessionEventCallbackTypes::OnSessionEnded>(
+        sessionId, callbacks, [](auto&& cbs) {
+            return cbs.OnSessionEnded;
+        },
+        [this](uint32_t sessionId, auto&& cb) {
+            auto token = std::make_shared<::uwb::OnSessionEndedToken>(sessionId, cb);
+            InsertSessionToken(m_onSessionEndedCallbacks, sessionId, token);
+            return token;
+        });
+
+    auto OnRangingStartedToken = GetToken<::uwb::UwbRegisteredSessionEventCallbackTypes::OnRangingStarted>(
+        sessionId, callbacks, [](auto&& cbs) {
+            return cbs.OnRangingStarted;
+        },
+        [this](uint32_t sessionId, auto&& cb) {
+            auto token = std::make_shared<::uwb::OnRangingStartedToken>(sessionId, cb);
+            InsertSessionToken(m_onRangingStartedCallbacks, sessionId, token);
+            return token;
+        });
+
+    auto OnRangingStoppedToken = GetToken<::uwb::UwbRegisteredSessionEventCallbackTypes::OnRangingStopped>(
+        sessionId, callbacks, [](auto&& cbs) {
+            return cbs.OnRangingStopped;
+        },
+        [this](uint32_t sessionId, auto&& cb) {
+            auto token = std::make_shared<::uwb::OnRangingStoppedToken>(sessionId, cb);
+            InsertSessionToken(m_onRangingStoppedCallbacks, sessionId, token);
+            return token;
+        });
+
+    auto OnPeerPropertiesChangedToken = GetToken<::uwb::UwbRegisteredSessionEventCallbackTypes::OnPeerPropertiesChanged>(
+        sessionId, callbacks, [](auto&& cbs) {
+            return cbs.OnPeerPropertiesChanged;
+        },
+        [this](uint32_t sessionId, auto&& cb) {
+            auto token = std::make_shared<::uwb::OnPeerPropertiesChangedToken>(sessionId, cb);
+            InsertSessionToken(m_onPeerPropertiesChangedCallbacks, sessionId, token);
+            return token;
+        });
+    auto OnSessionMembershipChangedToken = GetToken<::uwb::UwbRegisteredSessionEventCallbackTypes::OnSessionMembershipChanged>(
+        sessionId, callbacks, [](auto&& cbs) {
+            return cbs.OnSessionMembershipChanged;
+        },
+        [this](uint32_t sessionId, auto&& cb) {
+            auto token = std::make_shared<::uwb::OnSessionMembershipChangedToken>(sessionId, cb);
+            InsertSessionToken(m_onSessionMembershipChangedCallbacks, sessionId, token);
+            return token;
+        });
+
+    return {
+        OnSessionEndedToken,
+        OnRangingStartedToken,
+        OnRangingStoppedToken,
+        OnPeerPropertiesChangedToken,
+        OnSessionMembershipChangedToken
+    };
 }
 
 bool
 UwbConnector::CallbacksPresent()
 {
-    return (not m_sessionEventCallbacks.empty()) or (not m_deviceEventCallbacks.empty());
+    return not(m_onSessionEndedCallbacks.empty() and m_onRangingStartedCallbacks.empty() and

```

Comment:
Could use `std::any_of`:

`return std::any_of(std::cbegin({a, b, c, d, e, f, g, h}), std::cend({a, b, c, d, e, f, g, h}), [](const auto& callback) { 
return !callback.empty(); 
});`

Reply:
I don't think I can use any_of here because unfortunately the vectors have different types

Reply:
I actually think it might work still since they all have a `.empty()` method and are all iterable with `std::cbegin`. I'm definitely not positive so I could be wrong, but it might be worth a try. The current code is just a bit difficult to understand the logic and using `std::any_of` would make it clearer

Reply:
I tried it and it wouldn't compile. I believe it doesn't compile because std::any_of uses an iterator over whatever container its checking. However, the type of the iterator is ill-defined since the vectors are distinct types. It _might_ be able to work if I construct new vectors which just cast the members of the original vectors to RegisteredCallbackToken, but I don't want to incur that additional cost. Basically the problem is that the tokens like OnSessionEndedToken are derived classes of RegisteredCallbackToken but std::vector<OnSessionEndedToken> is not a derived class of std::vector<RegisteredCallbackToken>

Reply:
Ohh I see, well thanks for giving it a try!

Reply:
I think the types of each container being different is the limiting factor. `std::any_of` (and `std::ranges::any_of`) require a predicate which takes a certain type which is applied to each element of the input. I don't know how multiple vectors of different types could be adapted to take the same predicate.

@corbin-phipps There's certainly a way to do it with a traits-based approach as you described since all of the containers have an `empty()` function, but I don't see how it could be done with the STL constrained algorithms generically because you still need a way to group all of your inputs; the only way I know to do that is by combing the calls manually with `std::any_of(container<T>, pred<T>) && std::any_of(container<U>, pred<U>) && ...`.

You could do that combination elegantly with a template and C++ fold expressions, but I'm not sure it's worth the complexity. Something like (not sure if syntax is exactly correct here):

```
template <typename ContainerT, typename... Remaining>
requires <logic to ensure ContainerT has empty() function>
bool
AnyAreEmpty(const ContainerT& container, Remaining&&.. remainingContainers) {
    return (container.empty() || ...);
}
```

Reply:
wow, I had no idea that C++ fold expressions existed. Great to know for the future in case we might need them!